src/AWS/Lambda/Events/Kafka.hs

```diff
@@ -0,0 +1,249 @@
+
+keyFromText :: Text -> Key
+keyFromText = Key.fromText
+#else
+import qualified Data.HashMap.Strict    as KM
+type Key = Text
+
+keyToText :: Key -> Text
+keyToText = id
+
+keyFromText :: Text -> Key
+keyFromText = id
+#endif
+
+-- | Represents an event from either Amazon MSK or a self-managed
+-- Apache Kafka cluster, as the payloads are very similar.
+--
+-- The 'ToJSON' and 'FromJSON' instances on 'Record' perform base64
+-- conversion for you.
+--
+-- See the <https://docs.aws.amazon.com/lambda/latest/dg/with-kafka.html AWS documentation>
+-- for a sample payload.
+data KafkaEvent = KafkaEvent {
+  eventSource      :: !EventSource,
+  -- | Only present when using the "Amazon MSK" event source mapping.
+  eventSourceArn   :: !(Maybe Text),
+  bootstrapServers :: !(NonEmpty Text),
+  -- | The map's keys look like @"${topicName}-${partitionNumber}"@
+  records          :: !(Map Text [Record])
+} deriving (Eq, Show, Generic)
+
+instance FromJSON KafkaEvent where
+  parseJSON = withObject "KafkaEvent" $ \o ->
+    let parseBootstrapServers
+          = maybe (fail "bootstrapServers: empty string") pure
+          . NE.nonEmpty
+          . T.splitOn ","
+    in KafkaEvent
+         <$> o .: "eventSource"
+         <*> o .:! "eventSourceArn"
+         <*> (o .: "bootstrapServers" >>= parseBootstrapServers)
+         <*> o .: "records"
+
+instance ToJSON KafkaEvent where
+  toJSON e = object $
+    [ "eventSource" .= eventSource e
+    , "bootstrapServers" .= T.intercalate "," (NE.toList (bootstrapServers e))
+    , "records" .= records e
+    ] <> maybeToList (("eventSourceArn" .=) <$> eventSourceArn e)
+
+data EventSource
+  = AwsKafka -- ^ @"aws:kafka"@
+  | SelfManagedKafka -- ^ @"SelfManagedKafka"@
+  deriving (Eq, Show, Bounded, Enum, Ord, Generic)
+
+instance FromJSON EventSource where
+  parseJSON = withText "EventSource" $ \case
+    "aws:kafka" -> pure AwsKafka
+    "SelfManagedKafka" -> pure SelfManagedKafka
+    t -> fail $ "unrecognised EventSource: \"" <> show t <> "\""
+
+instance ToJSON EventSource where
+  toJSON = \case
+    AwsKafka -> String "aws:kafka"
+    SelfManagedKafka -> String "SelfManagedKafka"
+
+-- | Convenience alias: most of the time you will parse the records
+-- straight into some app-specific structure.
+type Record = Record' ByteString
+
+-- | Records from a Kafka event. This is 'Traversable', which means
+-- you can do things like parse a JSON-encoded payload:
+--
+-- @
+-- 'traverse' 'decodeStrict' :: 'FromJSON' a => Record ByteString -> Maybe (Record a)
+-- @
+data Record' a = Record {
+  topic :: !Text,
+  partition :: !Int32,
+  offset :: !Int64,
+  timestamp :: !Timestamp,
+  -- | NOTE: there can be multiple headers for a given key.
+  headers :: !(Vector Header),
+  key :: !(Maybe ByteString),
+  value :: !(Maybe a)
+} deriving (Eq, Show, Generic, Functor, Foldable, Traversable)
+
+-- | Decodes base64-encoded keys and values, where present.
+instance a ~ ByteString => FromJSON (Record' a) where
```

Comment:
This is to prevent `FromJSON`/`ToJSON` instances for other `Record' Foo` types, and force us to deserialise the stuff AWS sends us, and then inspect the payload later.

Reply:
So does this allow users to still write things like:  `instance FromJSON a => FromJSON (Record a)` without overlapping?  I'm not sure I understand the difference between this and `FromJSON (Record ByteString)`.

Reply:
The reverse, actually. When instance resolution happens, if type inference requires a `FromJSON` instance for `Record' Foo`, it will match this one, add the `Foo ~ ByteString` constraint, and then fail with a type error. This means that any (de)serialisation must first go through this instance, which ensures that we correctly base64 (en,de)code the payloads. People who know what they're doing can still provide an overlapping instance if they really want, or we can just provide `instance FromJSON (Record' ByteString)` and assume that people writing new instances know what they're doing.

(I might do this anyway - the use of the type equality in the context has been a continual source of comments among colleagues etc who I have shown it to.)